internal/pkg/agent/cmd/upgrade.go

```diff
@@ -53,20 +57,85 @@ func newUpgradeCommandWithArgs(_ []string, streams *cli.IOStreams) *cobra.Comman
 	cmd.Flags().String(flagPGPBytes, "", "PGP to use for package verification")
 	cmd.Flags().String(flagPGPBytesURI, "", "Path to a web location containing PGP to use for package verification")
 	cmd.Flags().String(flagPGPBytesPath, "", "Path to a file containing PGP to use for package verification")
+	cmd.Flags().BoolP(flagForce, "", false, "Advanced option to force an upgrade on a fleet managed agent")
+	err := cmd.Flags().MarkHidden(flagForce)
+	if err != nil {
+		log.Fatal(err)
+	}
 
 	return cmd
 }
 
+type upgradeInput struct {
+	streams   *cli.IOStreams
+	cmd       *cobra.Command
+	args      []string
+	c         client.Client
+	agentInfo info.Agent
+	cFunc     confirmFunc
+}
+
+type confirmFunc func(string, bool) (bool, error)
+
 func upgradeCmd(streams *cli.IOStreams, cmd *cobra.Command, args []string) error {
 	c := client.New()
-	return upgradeCmdWithClient(streams, cmd, args, c)
+	agentInfo, err := info.NewAgentInfoWithLog(cmd.Context(), "error", false)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve agent info while tring to upgrade the agent: %w", err)
+	}
+	input := &upgradeInput{
+		streams,
+		cmd,
+		args,
+		c,
+		agentInfo,
+		cli.Confirm,
+	}
+	return upgradeCmdWithClient(input)
+}
+
+func shouldUpgrade(cmd *cobra.Command, agentInfo info.Agent, cFunc confirmFunc) (bool, error) {
+	if agentInfo.IsStandalone() {
+		return true, nil
+	}
+
+	if agentInfo.Unprivileged() {
+		return false, fmt.Errorf("upgrade command needs to be executed as root for fleet managed agents")
+	}
+
+	force, err := cmd.Flags().GetBool(flagForce)
+	if err != nil {
+		return false, fmt.Errorf("failed to retrieve command flag information while trying to upgrade the agent: %w", err)
+	}
+
+	if !force {
+		return false, fmt.Errorf("upgrading fleet managed agents is not supported")
+	}
+
+	cf, err := cFunc("Upgrading fleet managed agents is not supported. Would you still like to proceed?", false)
+	if err != nil {
+		return false, fmt.Errorf("failed while confirming action: %w", err)
+	}
+
+	if !cf {
+		return false, fmt.Errorf("upgrade not confirmed")
```

Comment:
Please define a package variable with this error, using `errors.New`.
Here we are not formatting a new error, wrapping an inner error or some variable values in the error message so the call to `fmt.Errorf` is superfluous.

Bonus: we can assert the error in the test using `errors.Is` or its equivalent assertion function